compiler/src/dotty/tools/dotc/typer/VarianceChecker.scala

```diff
@@ -86,7 +86,15 @@ class VarianceChecker()(implicit ctx: Context) {
         case tp: TypeRef =>
           val sym = tp.symbol
           if (sym.variance != 0 && base.isContainedIn(sym.owner)) checkVarianceOfSymbol(sym)
-          else if (sym.isAliasType) this(status, sym.info.bounds.hi)
+          else if (sym.isAliasType) {
+            val aliasUpperBound = sym.info.bounds.hi
+            aliasUpperBound match {
+              case aliasUpperBound: RefinedType if aliasUpperBound.refinedInfo == tp =>
+                ctx.error("illegal cyclic reference involving " + sym + " and " + aliasUpperBound.refinedName, sym.pos)
```

Comment:
Even if this happens to explode in VarianceChecker right now, this isn't specific to VarianceChecker in any way. I would expect this to be checked in CheckNonCyclicMap